README.md
# kkb-epay

[![NPM version](https://img.shields.io/npm/v/kkb-epay.svg)](https://www.npmjs.com/package/kkb-epay)
[![npm](https://img.shields.io/npm/dt/kkb-epay.svg)](https://www.npmjs.com/package/kkb-epay)


ePay documentation can be found here https://testpay.kkb.kz/doc/htm/

## License

MIT


CHANGELOG.md
## master

## 0.0.1 (February 02, 2017)
- First commit


src/__tests__/client-test.js
/* @flow */
/* eslint-disable no-new */

import path from 'path';
import KkbEpayClient from '../kkbEpayClient';
import testConfig from '../../test_keys/config';

describe('KkbEpayClient', () => {
  const client = new KkbEpayClient(testConfig);
  const clientSelfSigned = new KkbEpayClient({
    ...testConfig,
    kkbcaPub: testConfig.certPub,
  });

  describe('constructor()', () => {
    it('should throw if opt.merchantId not provided', () => {
      expect(() => {
        new KkbEpayClient({
          ...testConfig,
          merchantId: undefined,
        });
      }).toThrowError(/should provide `merchantId`/);
    });

    it('should throw if opt.merchantName not provided', () => {
      expect(() => {
        new KkbEpayClient({
          ...testConfig,
          merchantName: undefined,
        });
      }).toThrowError(/should provide `merchantName`/);
    });

    it('should throw if opt.merchantCertificateId not provided', () => {
      expect(() => {
        new KkbEpayClient({
          ...testConfig,
          merchantCertificateId: undefined,
        });
      }).toThrowError(/should provide `merchantCertificateId`/);
    });

    it('should throw if opt.certPrvPass not provided', () => {
      expect(() => {
        new KkbEpayClient({
          ...testConfig,
          certPrvPass: undefined,
        });
      }).toThrowError(/should provide `certPrvPass`/);
    });

    it('should throw if opt.certPrv not provided', () => {
      expect(() => {
        new KkbEpayClient({
          ...testConfig,
          certPrv: undefined,
        });
      }).toThrowError(/should provide `certPrv`/);
    });

    it('should throw if opt.certPub not provided', () => {
      expect(() => {
        new KkbEpayClient({
          ...testConfig,
          certPub: undefined,
        });
      }).toThrowError(/should provide `certPub`/);
    });

    it('should throw if opt.kkbcaPub not provided', () => {
      expect(() => {
        new KkbEpayClient({
          ...testConfig,
          kkbcaPub: undefined,
        });
      }).toThrowError(/should provide `kkbcaPub`/);
    });
  });

  describe('_readFile()', () => {
    it('should return Promise', () => {
      const p = client._readFile('file');
      expect(p).toBeInstanceOf(Promise);
      return p.catch(() => {});
    });

    it('should reject if file is not provided', () => {
      // $FlowFixMe
      return client._readFile().catch(e => {
        expect(e).toBeDefined();
      });
    });

    it('should resolve with file content', async () => {
      const data = await client._readFile(path.resolve(__dirname, '../__mocks__/readFileTest.txt'));
      expect(data).toMatchSnapshot();
    });
  });

  describe('_sign()', () => {
    it('should return Promise', () => {
      const p = client._sign('test');
      expect(p).toBeInstanceOf(Promise);
      return p.catch(() => {});
    });

    it('should reject if provided empty data', () => {
      // $FlowFixMe
      return client._sign().catch(e => {
        expect(e.message).toMatch(/string.*buffer/);
      });
    });

    it('should reject if provided wrong passphrase', () => {
      const client2 = new KkbEpayClient({
        ...testConfig,
        certPrvPass: 'wrongPassphrase',
      });
      return client2._sign('Test').catch(e => {
        expect(e).toBeDefined();
      });
    });

    it('should resolve with signed data string', async () => {
      const data = await client._sign('Test');
      expect(data).toMatchSnapshot();
    });

    it('should invert sign if opts.invertSign = true', async () => {
      const client2 = new KkbEpayClient({
        ...testConfig,
        invertSign: false,
      });
      const client3 = new KkbEpayClient({
        ...testConfig,
        invertSign: true,
      });
      const data = 'Test';
      const sign = await client2._sign(data);
      const signInverted = await client3._sign(data);
      expect(sign).not.toEqual(signInverted);

      const signBuffer = Buffer.from(sign, 'base64');
      signBuffer.reverse();
      expect(signBuffer.toString('base64')).toEqual(signInverted);
    });
  });

  describe('_verify()', () => {
    it('should return Promise', () => {
      const p = clientSelfSigned._verify('test', 'sign');
      expect(p).toBeInstanceOf(Promise);
      return p.catch(() => {});
    });

    it('should reject if provided empty data', () => {
      // $FlowFixMe
      return clientSelfSigned._verify().catch(e => {
        expect(e.message).toMatch(/string.*buffer/);
      });
    });

    it('should resolve with false if provided wrong sign', () => {
      const data = 'Test';
      const sign = 'wrongSign';
      return clientSelfSigned._verify(data, sign).then(result => {
        expect(result).toBeFalsy();
      });
    });

    it('should resolve with true if provided correct sign', async () => {
      const data = 'Test';
      const sign = await clientSelfSigned._sign(data);
      await clientSelfSigned._verify(data, sign).then(result => {
        expect(result).toBeTruthy();
      });
    });
  });

  describe('_parseXml()', () => {
    it('should return Promise', () => {
      const p = client._parseXml('test');
      expect(p).toBeInstanceOf(Promise);
      return p.catch(() => {});
    });

    it('should reject with error on empty xmlString', () => {
      // $FlowFixMe
      return client._parseXml().catch(e => {
        expect(e.message).toContain('must be non-empty string');
      });
    });

    it('should resolve with js object', async () => {
      const res = await client._parseXml('<document><bank name="Kazkom"></bank></document>');
      expect(res).toEqual({
        document: {
          bank: [{ $: { name: 'Kazkom' } }],
        },
      });
    });
  });

  describe('_parseResponse()', () => {
    it('should return Promise', () => {
      const p = client._parseBankResponse('test');
      expect(p).toBeInstanceOf(Promise);
      return p.catch(() => {});
    });

    it('should reject with error on empty argument', () => {
      // $FlowFixMe
      return client._parseBankResponse().catch(e => {
        expect(e).toBeDefined();
      });
    });

    it('should reject with error on empty <bank_sign>', () => {
      return client
        ._parseBankResponse('<document><bank name="Kazkom"></bank></document>')
        .catch(e => {
          expect(e.message).toContain('should have non-empty `bank_sign`');
        });
    });

    it('should reject with error invalid <bank_sign>', async () => {
      const xml = await client._readFile(
        path.resolve(__dirname, '../__mocks__/testResponseInvalidSign.txt'),
      );

      try {
        await client._parseBankResponse(xml.trim());
      } catch (e) {
        expect(e.message).toContain('has unverified/wrong `bank_sign`');
      }
    });

    it('should resolve with <bank> data', async () => {
      const xml = await client._readFile(path.resolve(__dirname, '../__mocks__/testResponse.txt'));
      const result = await client._parseBankResponse(xml.trim());
      expect(result).toBeDefined();
    });
  });

  describe('_beautifyResponse()', () => {
    it('should remove arrays', () => {
      expect(client._beautifyResponse([{ body: 1 }])).toEqual({ body: 1 });

      expect(client._beautifyResponse([{ body: 1 }, { body: 2 }])).toEqual({
        body: 1,
        '1': { body: 2 },
      });
    });

    it('should expand args under $ key', () => {
      expect(client._beautifyResponse({ $: { arg1: 1, arg2: 2 } })).toEqual({
        arg1: 1,
        arg2: 2,
      });
    });

    it('should return regular keys', () => {
      expect(client._beautifyResponse({ key1: 1, key2: 2 })).toEqual({
        key1: 1,
        key2: 2,
      });
    });

    it('should return combined keys and args', () => {
      expect(client._beautifyResponse({ key1: 1, key2: 2, $: { arg1: 1, arg2: 2 } })).toEqual({
        key1: 1,
        key2: 2,
        arg1: 1,
        arg2: 2,
      });
    });

    it('should remove arrays in regular keys', () => {
      expect(client._beautifyResponse({ key1: [{ subKey1: 11, subKey2: 12 }], key2: 2 })).toEqual({
        key1: {
          subKey1: 11,
          subKey2: 12,
        },
        key2: 2,
      });
    });

    it('should match success response', async () => {
      const xml = await client._readFile(path.resolve(__dirname, '../__mocks__/testResponse.txt'));
      const result = await client._parseBankResponse(xml.trim());
      expect(client._beautifyResponse(result)).toEqual({
        name: 'Kazkommertsbank JSC',
        customer: {
          name: 'test',
          mail: 'SeFrolov@kkb.kz',
          phone: '+333333333',
          merchant: {
            cert_id: '00c183d70b',
            name: 'New Demo Shop',
            order: {
              amount: '10',
              currency: '398',
              department: { amount: '10', merchant_id: '92061103' },
              order_id: '0202171211',
            },
          },
          merchant_sign: { type: 'RSA' },
        },
        customer_sign: { type: 'RSA' },
        results: {
          timestamp: '2017-02-02 17:13:03',
          payment: {
            merchant_id: '92061103',
            card: '440564-XX-XXXX-6150',
            amount: '10',
            reference: '170202171303',
            approval_code: '171303',
            response_code: '00',
            Secure: 'No',
            card_bin: '',
            c_hash: '13988BBF7C6649F799F36A4808490A3E',
          },
        },
      });
    });

    it('should match error response', async () => {
      const xml = await client._readFile(
        path.resolve(__dirname, '../__mocks__/testResponseError.txt'),
      );
      const result = await client._parseXml(xml);
      expect(client._beautifyResponse(result)).toEqual({
        response: {
          error: {
            _: 'Error Message',
            code: '00',
            time: '2006-11-22 12:20:30',
            type: 'system | auth',
          },
          order_id: '123456',
          session: { id: '1234654656545' },
        },
      });
    });
  });

  describe('_createXml()', () => {
    it('should return string', () => {
      const str = client._createXml('root', { a: 1, b: 2 });
      expect(str).toEqual('<root><a>1</a><b>2</b></root>');
    });

    it('should accept args', () => {
      const str = client._createXml('root', { $: { a: 1, b: 2 }, t: 3 });
      expect(str).toEqual('<root a="1" b="2"><t>3</t></root>');
    });
  });

  describe('_createOrderXML()', () => {
    it('should create signed doc', async () => {
      const signedOrder = await client._createOrderXML('000333', 500, 398);
      expect(signedOrder).toMatchSnapshot();

      // $FlowFixMe
      const marchantData = signedOrder.match(/(<merchant.*<\/merchant>)/i)[1];
      // $FlowFixMe
      const marchantSign = signedOrder.match(/<merchant_sign.*>(.+)<\/merchant_sign>/i)[1];
      expect(marchantData).toMatchSnapshot();
      expect(marchantSign).toMatchSnapshot();
      expect(await clientSelfSigned._verify(marchantData, marchantSign)).toBeTruthy();
    });
  });

  describe('createOrder()', () => {
    const createOrderOpts = {
      orderId: '000001',
      amount: 500,
      currency: 398,
      callbackUrl: '/process',
      successUrl: '/success',
      failureUrl: '/failure',
      email: '',
      ln: 'rus',
    };

    it('should check orderId option', () => {
      return client
        .createOrder({
          ...createOrderOpts,
          orderId: '1',
        })
        .catch(e => {
          expect(e.message).toContain('minLength 6 symbols and maxLength 15');
        });
    });

    it('should check amount option', () => {
      return client
        .createOrder({
          ...createOrderOpts,
          amount: 0,
        })
        .catch(e => {
          expect(e.message).toContain('`opts.amount` as positive number');
        });
    });

    it('should check callbackUrl option', () => {
      return client
        .createOrder({
          ...createOrderOpts,
          callbackUrl: '',
        })
        .catch(e => {
          expect(e.message).toContain('should provide `opts.callbackUrl`');
        });
    });

    it('should check successUrl option', () => {
      return client
        .createOrder({
          ...createOrderOpts,
          successUrl: '',
        })
        .catch(e => {
          expect(e.message).toContain('should provide `opts.successUrl`');
        });
    });

    it('should check failureUrl option', () => {
      return client
        .createOrder({
          ...createOrderOpts,
          failureUrl: '',
        })
        .catch(e => {
          expect(e.message).toContain('should provide `opts.failureUrl`');
        });
    });

    it('should resolve valid result', async () => {
      const res = await client.createOrder(createOrderOpts);
      expect(res).toEqual({
        BackLink: '/success',
        FailureBackLink: '/failure',
        Language: 'rus',
        PostLink: '/process',
        Signed_Order_B64: 'PGRvY3VtZW50PjxtZXJjaGFudCBjZXJ0X2lkPSIwMEMxODJCMTg5IiBuYW1lPSJUZXN0IHNob3AiPjxvcmRlciBvcmRlcl9pZD0iMDAwMDAxIiBhbW91bnQ9IjUwMCIgY3VycmVuY3k9IjM5OCI+PGRlcGFydG1lbnQgbWVyY2hhbnRfaWQ9IjkyMDYxMTAxIiBhbW91bnQ9IjUwMCIvPjwvb3JkZXI+PC9tZXJjaGFudD48bWVyY2hhbnRfc2lnbiB0eXBlPSJSU0EiPlBEcVZEMkpSNzhzUjI3SUxodjFvZk5GeFF5aVp0anQ3UHRFV2NORlBIRVZRODJGRlRCMVcyWmZuazhHcHdSMlF6dnRNb0U0QzlJbkVSTjlGWkRHMkR3PT08L21lcmNoYW50X3NpZ24+PC9kb2N1bWVudD4=',
        email: '',
      });
    });
  });

  describe('processResponseCreateOrder()', () => {
    it('should return Promise', () => {
      const p = client.processResponseCreateOrder('<xml />');
      expect(p).toBeInstanceOf(Promise);
      return p.catch(() => {});
    });

    it('should resolve with parsed data on valid response', async () => {
      const validResponseXml = await client._readFile(
        path.resolve(__dirname, '../__mocks__/testResponse.txt'),
      );

      const res = await client.processResponseCreateOrder(validResponseXml);
      expect(res).toEqual({
        timestamp: '2017-02-02 17:13:03',
        name: 'test',
        mail: 'SeFrolov@kkb.kz',
        phone: '+333333333',
        order_id: '0202171211',
        amount: '10',
        currency: '398',
        merchant_id: '92061103',
        card: '440564-XX-XXXX-6150',
        reference: '170202171303',
        approval_code: '171303',
        response_code: '00',
        Secure: 'No',
        card_bin: '',
        c_hash: '13988BBF7C6649F799F36A4808490A3E',
      });
    });

    it('should reject if error response from bank', () => {
      return client
        ._readFile(path.resolve(__dirname, '../__mocks__/testResponseError.txt'))
        .then(xml => client.processResponseCreateOrder(xml))
        .catch(e => {
          expect(e.message).toContain('Response should have non-empty `bank_sign` property.');
        });
    });

    it('should reject if invalid sign', () => {
      return client
        ._readFile(path.resolve(__dirname, '../__mocks__/testResponseInvalidSign.txt'))
        .then(xml => client.processResponseCreateOrder(xml))
        .catch(e => {
          expect(e.message).toContain('Response has unverified/wrong `bank_sign`');
        });
    });
  });

  describe('proceedOrder()', () => {
    it('should create signed proceed command', async () => {
      const signedCmd = await client.proceedOrder(
        'complete',
        'referenceNumber',
        'approvalCode',
        'orderId',
        500,
        398,
      );
      expect(signedCmd).toMatchSnapshot();
    });

    it('should has `reason` tag for reverse command', async () => {
      const signedCmd = await client.proceedOrder(
        'reverse',
        'referenceNumber',
        'approvalCode',
        'orderId',
        500,
        398,
      );
      expect(signedCmd).toContain('<reason>Return payment</reason>');
      expect(signedCmd).toMatchSnapshot();
    });
  });
});


flow-typed/npm/xml2js_vx.x.x.js
// flow-typed signature: c774495339982f354b31cf2b59f2f98a
// flow-typed version: <<STUB>>/xml2js_v^0.4.17/flow_v0.38.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   'xml2js'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the 
 * community by sending a pull request to: 
 * https://github.com/flowtype/flow-typed
 */

declare module 'xml2js' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module 'xml2js/lib/bom' {
  declare module.exports: any;
}

declare module 'xml2js/lib/processors' {
  declare module.exports: any;
}

declare module 'xml2js/lib/xml2js' {
  declare module.exports: any;
}

// Filename aliases
declare module 'xml2js/lib/bom.js' {
  declare module.exports: $Exports<'xml2js/lib/bom'>;
}
declare module 'xml2js/lib/processors.js' {
  declare module.exports: $Exports<'xml2js/lib/processors'>;
}
declare module 'xml2js/lib/xml2js.js' {
  declare module.exports: $Exports<'xml2js/lib/xml2js'>;
}


flow-typed/npm/jest-babel_vx.x.x.js
// flow-typed signature: 0af2efe8c8264b2b73afc46c4e9dea3d
// flow-typed version: <<STUB>>/jest-babel_v^1.0.1/flow_v0.38.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   'jest-babel'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the 
 * community by sending a pull request to: 
 * https://github.com/flowtype/flow-typed
 */

declare module 'jest-babel' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */


// Filename aliases
declare module 'jest-babel/index' {
  declare module.exports: $Exports<'jest-babel'>;
}
declare module 'jest-babel/index.js' {
  declare module.exports: $Exports<'jest-babel'>;
}


flow-typed/npm/object-path_vx.x.x.js
// flow-typed signature: 34982bd243234b3334149332397d8b4e
// flow-typed version: <<STUB>>/object-path_v^0.11.3/flow_v0.38.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   'object-path'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the 
 * community by sending a pull request to: 
 * https://github.com/flowtype/flow-typed
 */

declare module 'object-path' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module 'object-path/benchmark' {
  declare module.exports: any;
}

declare module 'object-path/test' {
  declare module.exports: any;
}

// Filename aliases
declare module 'object-path/benchmark.js' {
  declare module.exports: $Exports<'object-path/benchmark'>;
}
declare module 'object-path/index' {
  declare module.exports: $Exports<'object-path'>;
}
declare module 'object-path/index.js' {
  declare module.exports: $Exports<'object-path'>;
}
declare module 'object-path/test.js' {
  declare module.exports: $Exports<'object-path/test'>;
}


flow-typed/npm/eslint-config-prettier_vx.x.x.js
// flow-typed signature: 1b798012f9475480b79d5a932c5eee26
// flow-typed version: <<STUB>>/eslint-config-prettier_v^1.0.2/flow_v0.38.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   'eslint-config-prettier'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the 
 * community by sending a pull request to: 
 * https://github.com/flowtype/flow-typed
 */

declare module 'eslint-config-prettier' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module 'eslint-config-prettier/cli' {
  declare module.exports: any;
}

declare module 'eslint-config-prettier/rules/core' {
  declare module.exports: any;
}

declare module 'eslint-config-prettier/rules/react' {
  declare module.exports: any;
}

// Filename aliases
declare module 'eslint-config-prettier/cli.js' {
  declare module.exports: $Exports<'eslint-config-prettier/cli'>;
}
declare module 'eslint-config-prettier/index' {
  declare module.exports: $Exports<'eslint-config-prettier'>;
}
declare module 'eslint-config-prettier/index.js' {
  declare module.exports: $Exports<'eslint-config-prettier'>;
}
declare module 'eslint-config-prettier/rules/core.js' {
  declare module.exports: $Exports<'eslint-config-prettier/rules/core'>;
}
declare module 'eslint-config-prettier/rules/react.js' {
  declare module.exports: $Exports<'eslint-config-prettier/rules/react'>;
}


src/kkbEpayClient.js
/* @flow */
/* eslint-disable class-methods-use-this */

import fs from 'fs';
import crypto from 'crypto';
import xml2js from 'xml2js';
import objectPath from 'object-path';

type absolutePathStringT = string;

export type KkbEpayClientOptsT = {
  merchantId: string,
  merchantName: string,
  merchantCertificateId: string,
  certPrv: absolutePathStringT,
  certPrvPass: string,
  certPub: absolutePathStringT,
  kkbcaPub: absolutePathStringT,
  invertSign?: boolean,
  serverEndpoint: string,
};

export type ProceedOrderCmdT = 'reverse' | 'complete' | 'refund';
export type CurrencyISO4217 = 398 | 840 | 643; // 398 - KZT, 840 - USD, 643 - RUB;

export default class KkbEpayClient {
  opts: KkbEpayClientOptsT;

  constructor(opts: KkbEpayClientOptsT) {
    if (!opts.merchantId) {
      throw new Error(
        'You should provide `merchantId` option (string). ' +
          'See provided by Qazkom variable MERCHANT_ID in `config.txt`',
      );
    }
    if (!opts.merchantName) {
      throw new Error(
        'You should provide `merchantName` option (string). ' +
          'See provided by Qazkom variable MERCHANT_NAME in `config.txt`',
      );
    }
    if (!opts.merchantCertificateId) {
      throw new Error(
        'You should provide `merchantCertificateId` option (string). ' +
          'See provided by Qazkom variable MERCHANT_CERTIFICATE_ID in `config.txt`',
      );
    }
    if (!opts.certPrvPass) {
      throw new Error(
        'You should provide `certPrvPass` option (string). ' +
          'See provided by Qazkom variable PRIVATE_KEY_PASS in `config.txt`',
      );
    }
    if (!opts.certPrv) {
      throw new Error(
        'You should provide `certPrv` option (absolute path). ' +
          'This file provided by Qazkom and has name `cert.prv`',
      );
    }
    if (!opts.certPub) {
      throw new Error(
        'You should provide `certPub` option (absolute path). ' +
          'This file provided by Qazkom and has name `cert.pub`',
      );
    }
    if (!opts.kkbcaPub) {
      throw new Error(
        'You should provide `kkbcaPub` option (absolute path). ' +
          'This file provided by Qazkom and has name `kkbca.pem` or `kkbca.pub`',
      );
    }

    this.opts = {
      ...opts,
      invertSign: (
        // true by default
        opts.invertSign !== false
      ),
      serverEndpoint: opts.serverEndpoint || 'https://epay.kkb.kz/',
    };
  }

  _readFile(
    filename: string,
    enc: buffer$Encoding = 'ascii', // eslint-disable-line
  ): Promise<string> {
    return new Promise((resolve, reject) => {
      try {
        fs.readFile(filename, (err, buffer) => {
          if (err) {
            reject(err);
          } else {
            resolve(buffer.toString(enc));
          }
        });
      } catch (err) {
        reject(err);
      }
    });
  }

  _sign(data: string): Promise<string> {
    return this._readFile(this.opts.certPrv).then(key => {
      const signer = crypto.createSign('RSA-SHA1');
      signer.update(data);
      const pk = {
        key,
        passphrase: this.opts.certPrvPass,
      };

      const signBuffer = signer.sign(pk);

      if (this.opts.invertSign) {
        signBuffer.reverse();
      }

      return signBuffer.toString('base64');
    });
  }

  _verify(data: string, sign: string): Promise<boolean> {
    return this._readFile(this.opts.kkbcaPub).then(key => {
      const verify = crypto.createVerify('RSA-SHA1');
      verify.update(data);

      const signBuffer = Buffer.from(sign, 'base64');
      if (this.opts.invertSign) {
        signBuffer.reverse();
      }
      const res = verify.verify(key, signBuffer);
      return res;
    });
  }

  _createXml(rootName: string, obj: Object): string {
    const builder = new xml2js.Builder({
      rootName,
      headless: true,
      renderOpts: {
        pretty: false,
      },
    });
    return builder.buildObject(obj);
  }

  _parseXml(xml: string): Promise<Object> {
    return new Promise((resolve, reject) => {
      if (!xml || typeof xml !== 'string') {
        reject(new Error('Argument must be non-empty string.'));
        return;
      }

      const parser = new xml2js.Parser();
      parser.parseString(xml, (err, data) => {
        if (err) {
          reject(err);
          return;
        }
        resolve(data);
      });
    });
  }

  async _parseBankResponse(xml: string): Promise<Object> {
    if (!xml) {
      throw new Error('Response should be non-empty string.');
    }

    const res = await this._parseXml(xml);

    const signObj = objectPath.get(res, 'document.bank_sign.0');
    if (!signObj || !signObj._) {
      throw new Error('Response should have non-empty `bank_sign` property.');
    }

    const sign = signObj._;
    const bankMatch = xml.match(/(<bank\s.*<\/bank>)/i);
    const isValid = bankMatch && (await this._verify(bankMatch[1], sign));
    if (!isValid) {
      throw new Error(
        'Response has unverified/wrong `bank_sign`. ' +
          'It may be unauthorized request. ' +
          'Or was provided wrong bank public key in `opts.kkbcaPub` property.',
      );
    }

    return objectPath.get(res, 'document.bank');
  }

  _beautifyResponse(xmlObj: any): Object {
    let res = {};
    if (Array.isArray(xmlObj)) {
      if (xmlObj.length > 0) {
        res = this._beautifyResponse(xmlObj[0]);
        for (let i = 1; i < xmlObj.length; i++) {
          res[i.toString()] = this._beautifyResponse(xmlObj[i]);
        }
      }
    } else if (typeof xmlObj === 'object') {
      Object.keys(xmlObj).forEach(key => {
        if (key === '$') {
          res = { ...res, ...xmlObj.$ };
        } else if (Array.isArray(xmlObj[key]) || typeof xmlObj[key] === 'object') {
          res[key] = this._beautifyResponse(xmlObj[key]);
        } else {
          res[key] = xmlObj[key];
        }
      });
    }
    return res;
  }

  async _createOrderXML(
    orderId: string,
    amount: number,
    currency: CurrencyISO4217 = 398,
  ): Promise<string> {
    const merchantObj = {
      $: {
        cert_id: this.opts.merchantCertificateId,
        name: this.opts.merchantName,
      },
      order: [
        {
          $: {
            order_id: orderId,
            amount,
            currency,
          },
          department: [
            {
              $: {
                merchant_id: this.opts.merchantId,
                amount,
              },
            },
          ],
        },
      ],
    };

    const merchantTag = this._createXml('merchant', merchantObj);
    const merchantSign = await this._sign(merchantTag);
    return `<document>${merchantTag}<merchant_sign type="RSA">${merchantSign}</merchant_sign></document>`;
  }

  getCreateOrderUrl() {
    return `${this.opts.serverEndpoint}/jsp/process/logon.jsp`;
  }

  getProceedOrderUrl() {
    return `${this.opts.serverEndpoint}/jsp/remote/control.jsp`;
  }

  async createOrder(
    opts: {
      orderId: string,
      amount: number,
      currency: CurrencyISO4217,
      email: string,
      callbackUrl: string,
      successUrl: string,
      failureUrl: string,
      ln: 'rus' | 'eng' | 'kaz',
    },
  ): Promise<Object> {
    if (!opts) {
      throw new Error('You provide empty options');
    }

    if (
      !opts.orderId ||
        typeof opts.orderId !== 'string' ||
        opts.orderId.length < 6 ||
        opts.orderId.length > 15
    ) {
      throw new Error(
        'You should provide `opts.orderId` as string with minLength 6 symbols and maxLength 15.',
      );
    }

    if (!opts.amount || opts.amount < 0) {
      throw new Error('You should provide `opts.amount` as positive number.');
    }

    if (!opts.callbackUrl) {
      throw new Error(
        'You should provide `opts.callbackUrl`. Allowed ports 80, 443. On this urls bank server will send you response.',
      );
    }

    if (!opts.successUrl) {
      throw new Error(
        'You should provide `opts.successUrl`. Client will be redirected on this url after success payment.',
      );
    }

    if (!opts.failureUrl) {
      throw new Error(
        'You should provide `opts.failureUrl`. Client will be redirected on this url after failure payment.',
      );
    }

    const signedOrder = await this._createOrderXML(opts.orderId, opts.amount, opts.currency);
    const signedOrderBuffer = new Buffer(signedOrder);

    return {
      email: opts.email || '',
      // Signed_Order: signedOrder, // FOR DEBUG PURPOSES
      Signed_Order_B64: signedOrderBuffer.toString('base64'),
      BackLink: opts.successUrl,
      FailureBackLink: opts.failureUrl,
      PostLink: opts.callbackUrl,
      Language: opts.ln || 'rus',
    };
  }

  async processResponseCreateOrder(xml: string): Promise<Object> {
    try {
      const xmlObj = await this._parseBankResponse(xml);
      const res = this._beautifyResponse(xmlObj);

      const responseCode = objectPath.get(res, 'results.payment.response_code');
      if (responseCode !== '00') {
        return Promise.reject(new Error(
          `results.payment.response_code: ${responseCode} is not equal to 00`,
        ));
      }

      // const merchantId = objectPath.get(res, 'results.payment.merchant_id');
      // if (merchantId !== this.opts.merchantId) {
      //   return Promise.reject(new Error(
      //     `results.payment.merchant_id: ${merchantId} is not equal to your id in opts ${this.opts.merchantId}`
      //   ));
      // }

      return {
        timestamp: res.results.timestamp,
        name: res.customer.name,
        mail: res.customer.mail,
        phone: res.customer.phone,
        order_id: res.customer.merchant.order.order_id,
        amount: res.customer.merchant.order.amount,
        currency: res.customer.merchant.order.currency,
        ...res.results.payment,
      };
    } catch (e) {
      return Promise.reject(e);
    }
  }

  /**
  * cmd 'reverse' - return blocked amount back to customer
  *     'complete' - confirm payment and transfer blocket amount to merchant
  *     'refund' - return confirmed payment back
  */
  async proceedOrder(
    cmd: ProceedOrderCmdT,
    reference: string,
    approvalCode: string,
    orderId: string,
    amount: number,
    currency: CurrencyISO4217,
  ): Promise<string> {
    const merchantObj = {
      $: {
        merchant_id: this.opts.merchantId,
      },
      command: {
        $: {
          type: cmd,
        },
      },
      payment: {
        $: {
          reference,
          approval_code: approvalCode,
          orderid: orderId,
          amount,
          currency_code: currency,
        },
      },
    };

    if (cmd === 'reverse') {
      // $FlowFixMe
      merchantObj.reason = 'Return payment';
    }

    const merchantTag = this._createXml('merchant', merchantObj);
    const merchantSign = await this._sign(merchantTag);
    return `<document>${merchantTag}<merchant_sign type="RSA" cert_id="${this.opts.merchantCertificateId}">${merchantSign}</merchant_sign></document>`;
  }

  async processResponseProceedOrder(xml: string): Promise<Object> {
    try {
      const xmlObj = await this._parseBankResponse(xml);
      const res = this._beautifyResponse(xmlObj);

      if (objectPath.get(res, 'bank.response.code') !== '00') {
        return Promise.reject(new Error('Response.code is not equal to 00'));
      }

      return res;
    } catch (e) {
      return Promise.reject(e);
    }
  }
}
